### PR TITLE
CHIP-0005: NFT1 Standard

### PR DESCRIPTION
**Pull request by danieljperry**

### CODE REVIEW

File: CHIPs/chip-dan-perry-nft1.md [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+  * Each NFT may be owned by one DID, based on Chia's DID1 Standard
+  * If an NFT is minted using a DID, then the NFT's provenance must be traceable to the DID of the NFT's Minter
+  
+* **Permanence**
+  * Required: 3 URI lists must be included with each NFT, which map to:
+    1. The NFT's main content
+    2. The NFT's metadata
+    3. The NFT's licensing and legal requirements
+  * The Owner may append links as needed; the Owner may not remove links
+  * Upon minting an NFT, a SHA-256 hash of the file being linked to shall be embedded within the NFT itself. This gives a permanent way to verify the NFT's contents, even if each of the links is broken
```
Comment (BrandtH22) [inline CHIPs/chip-dan-perry-nft1.md]: Do we need to define that the hash is added separately to allow for delayed reveals? 
ex. provide the correct file hash but a URL that links to a dummy image until the NFT is revealed

File: CHIPs/chip-dan-perry-nft1.md [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+* NFT Owners must be able to view a list of NFTs they own. This includes both ownership scenarios for NFTs – those owned by a DID, and those owned by an XCH address
+* The Minter must include a Transfer Program with the NFT upon its minting. This program sets the standard governing how the NFT will be transferred throughout its life. The Transfer Program may not be modified after the NFT's minting
+* NFTs must support transaction fees upon gifting, selling, transferring or swapping NFTs. These fees are paid to Chia's farmers and are completely separate from royalty payments
+* Wallets that implement NFT1 must automatically discover new NFTs owned by the wallet’s owner – whether via a DID or an XCH address – and add them to the wallet
+* Users of wallets that implement NFT1 must be able to view a gallery of the NFTs they own
+* Users of wallets that implement NFT1 must be able to view the metadata and license information for all NFTs they own
+* NFT1 NFTs must be image files. The only supported types are jpg, gif, png and svg
```
Comment (BrandtH22) [inline CHIPs/chip-dan-perry-nft1.md]: We should not limit the file types. I propose changing must to should here.

File: CHIPs/chip-dan-perry-nft1.md [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+* **NFT Address:** An NFT's ID shall be its launcher, which is a bech32-encoded address with a prefix of "nft"
+* **Minter's DID:** If a DID was used in this NFT's minting, then the Minter’s DID shall be discoverable by blockchain explorers
+* **Data URI list:** A list of URIs that map to the NFT's main content
+  * This list shall be stored as bytes, which may be encoded in any format
+  * This list must contain at least one URI
+  * The Owner optionally may prepend to this list, such that the newest URI will always be the list’s first element. The Owner shall not shorten or remove the list
+  * Each URI may be a:
+    * Decentralized storage link, such as IPFS, Arweave, Storj, etc
+    * Web URL
+  * Upon the NFT's minting, the content of the first item in this list must hash to the Data Hash (defined below)
```
Comment (BrandtH22) [inline CHIPs/chip-dan-perry-nft1.md]: I propose changing this must to a should.
See above comment regarding delayed reveals.

File: CHIPs/chip-dan-perry-nft1.md [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+* **Metadata URI list:** A list of URIs that map to the NFT's metadata
+  * This list shall be stored as bytes, which may be encoded in any format
+  * This list must contain at least one URI
+  * The Owner optionally may append to this list, but they shall not shorten or remove it
+  * This list must be curried into the NFT's puzzle
+  * Upon the NFT's minting and sale, the content of the first item in this list must hash to the Metadata Hash (defined below)
```
Comment (BrandtH22) [inline CHIPs/chip-dan-perry-nft1.md]: I propose changing this must to a should.
See above comment regarding delayed reveals.

File: CHIPs/chip-dan-perry-nft1.md [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+      * "id" is the string identifier of the collection
+    * **attributes**
+      * contains zero or more key/value pairs
+      * "trait_type" is the name of the trait
+      * "value" can be either a number or a string
+      * "min_value" (optional). This sets a floor for a numerical trait’s possible values. It may not be used for string values
+      * "max_value" is (optional). This sets a ceiling for a numerical trait's possible values. It may not be used for string values
+    * The following JSON format is required (other than recommended and optional fields):
+     ```json
+      {
```
Comment (BrandtH22) [inline CHIPs/chip-dan-perry-nft1.md]: Can we add a version field as proposed here:
https://developers.chia.net/t/nft-metadata-file-schema/433/15

File: CHIPs/chip-dan-perry-nft1.md [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+          {
+            "trait_type": "Sample type 3",
+            "value": 10,
+            "min_value": 4
+            "max_value": 11
+          }
+        ]
+      }
+      ```
+  * **License URI list:** A list of URIs that map to the NFT's licensing and legal requirements
```
Comment (BrandtH22) [inline CHIPs/chip-dan-perry-nft1.md]: Will there be recommended licenses upon NFT1 launch?

File: CHIPs/chip-dan-perry-nft1.md [hunk 7]
```diff
@@ -0,0 +1,409 @@
...
+
+The metadata updater program can also update itself by using this puzzle:
+[https://github.com/Chia-Network/chia-blockchain/blob/nft1_ownership_layer/chia/wallet/puzzles/nft_metadata_updater_updateable.clvm](https://github.com/Chia-Network/chia-blockchain/blob/nft1_ownership_layer/chia/wallet/puzzles/nft_metadata_updater_updateable.clvm)
+
+### Inner Puzzle
+This puzzle defines the primary characteristics of the NFT. It is located in GitHub, under [https://github.com/Chia-Network/chia-blockchain/blob/release/1.4.0/chia/wallet/puzzles/nft_state_layer.clvm](https://github.com/Chia-Network/chia-blockchain/blob/release/1.4.0/chia/wallet/puzzles/nft_state_layer.clvm).
+
+### Offer 2.0
+The NFT1 standard includes Offer 2.0, which builds on top of the [Offer 1.0 standard](https://chialisp.com/docs/puzzles/offers) to enable the buying, selling and trading of XCH, CATs and NFTs.
```
Comment (BrandtH22) [inline CHIPs/chip-dan-perry-nft1.md]: Will offer2 have its own CHIP?

File: CHIPs/chip-dan-perry-nft1.md [hunk 8]
```diff
@@ -0,0 +1,409 @@
...
+* Must specify the NFT ID of any NFTs they want to offer
+* Must be shown a visual image of all NFTs they are offering to send and receive
+* May see their Open and Accepted NFT offers, including the creation date of all offers
+
+**Takers**
+* Must be shown all royalties associated with all NFTs being offered
+* Must be shown a visual image of any NFTs they will send and receive in the offer
+* May see their Open and Accepted NFT offers, including creation date of all offers
+
+### RPC Calls
```
Comment (BrandtH22) [inline CHIPs/chip-dan-perry-nft1.md]: These all appear to be the wallet RPC commands, will there be additional Full Node RPC commands for NFT1?

**Review by BrandtH22 — COMMENTED** [review]
Overall, the CHIP is looking great and I am excited for NFT1!

File: CHIPs/chip-dan-perry-nft1.md [hunk 9]
```diff
@@ -0,0 +1,409 @@
...
+        ]
+      }
+      ```
+  * **License URI list:** A list of URIs that map to the NFT's licensing and legal requirements
+    * This list shall be stored as bytes, which may be encoded in any format
+    * This list must contain at least one URI
+    * The Owner optionally may append to this list, but they shall not shorten or remove it
+    * Upon the NFT's minting, the content of the first item in this list must hash to the the License Hash (defined below)
+  * **Data Hash:** The SHA-256 hash of the NFT’s content. For example, if the NFT is a .jpeg file, then the Data Hash is the hash of that file. Upon the NFT's minting, this hash must match the hash of the content of the first item of the Data URI list (defined above). This hash shall not be modified after the NFT's minting
+  * **Metadata Hash:** A SHA-256 hash of the contents of the first item of the Metadata URI list. The Owner may modify this hash
```
Comment (BrandtH22) [inline CHIPs/chip-dan-perry-nft1.md]: If the owner can modify the metadata hash could they change the attributes that ascribe rarity of an NFT?

**Comment by CommanderMoto** [comment]
So, since this CHIP is in the wild as of v1.4.0 ... isn't it time we merge this PR for real?

**Comment by danieljperry** [comment]
> So, since this CHIP is in the wild as of v1.4.0 ... isn't it time we merge this PR for real?

The CHIP is the NFT1 standard. Chia version 1.4 is a separate thing. It implements most of this standard, but not quite all of it yet. Soon we'll go through the formal review process and make any last-minute changes before finalizing the standard. Once it's final it can't change, so we want to make sure we get it right.

File: CHIPs/chip-dan-perry-nft1.md (hunk 1 above)
Comment (danieljperry) [reply CHIPs/chip-dan-perry-nft1.md]: Agreed and updated.

File: CHIPs/chip-dan-perry-nft1.md (hunk 2 above)
Comment (danieljperry) [reply CHIPs/chip-dan-perry-nft1.md]: Updated to allow any media format.

File: CHIPs/chip-dan-perry-nft1.md (hunk 3 above)
Comment (danieljperry) [reply CHIPs/chip-dan-perry-nft1.md]: updated

File: CHIPs/chip-dan-perry-nft1.md (hunk 5 above)
Comment (danieljperry) [reply CHIPs/chip-dan-perry-nft1.md]: The metadata contents have been moved outside of this CHIP. See CHIP7 for more info.

File: CHIPs/chip-dan-perry-nft1.md (hunk 9 above)
Comment (danieljperry) [reply CHIPs/chip-dan-perry-nft1.md]: This has been updated to "This hash shall not be modified after the NFT's minting".

File: CHIPs/chip-dan-perry-nft1.md (hunk 8 above)
Comment (danieljperry) [reply CHIPs/chip-dan-perry-nft1.md]: There aren't any specific full node RPCs in the current NFT standard.

File: CHIPs/chip-0005.md [hunk 10]
```diff
@@ -0,0 +1,393 @@
...
+Custom Transfer Programs with different functionality optionally may be deployed. They shall be considered conformant with this CHIP.
+
+## Specification
+
+### NFT Structure
+Each NFT that follows the NFT1 standard is required to be implemented as a singleton. Because of this, users are required to spend a minimum of 1 mojo to create an NFT.
+
+Each NFT must contain the following information:
+* **NFT Address:** An NFT's ID shall be its launcher, which is a bech32-encoded address with a prefix of "nft"
+* **Minter's DID:** If a DID was used in this NFT's minting, then the Minter’s DID shall be discoverable by blockchain explorers
```
Comment (ddesiderio) [inline CHIPs/chip-0005.md:166]: Can you add more info on how this linking between NFT<->DID is done and can be verified? For example, if there's a stamp of a DID signature required to be added after each NFT transfer/mint how will that look and how can it be verified? Some examples may help for this too, to ensure that community tools snap to the same way of displaying/verifying provenance

Comment (danieljperry) [reply CHIPs/chip-0005.md]: This requirement has been removed.

File: CHIPs/chip-0005.md [hunk 11]
```diff
@@ -0,0 +1,392 @@
...
+    * Decentralized storage link, such as IPFS, Arweave, Storj, etc
+    * Web URL
+  * Upon the NFT's minting, the content of the first item in this list is recommmended to hash to the Data Hash (defined below)
+* **Metadata URI list:** A list of URIs that map to the NFT's metadata
+  * This list shall be stored as bytes, which may be encoded in any format
+  * This list must contain at least one URI
+  * The Owner optionally may prepend to this list, but they shall not modify or remove it
+  * This list must be curried into the NFT's puzzle
+  * Upon the NFT's minting and sale, the content of the first item in this list is recommmended to hash to the Metadata Hash (defined below)
+  * There will be multiple off-chain metadata formats, each of which will have its own Process CHIP. These CHIPs shall specify a list of required, recommended, and/or optional fields. If a formalised specification can be produced (such as a JSON schema dialect), then this should be included in the CHIP's assets directory to enable schema verification
```
Comment (wriches) [inline CHIPs/chip-0005.md:180]: JSON should be recommended but not mandatory for off-chain metadata formats, in order to maximise interoperability and reduce the logic required to process multiple formats.

Comment (wriches) [inline CHIPs/chip-0005.md:180]: Charset requirements should be provided here. Something like "Off-chain metadata formats should be encoded in UTF-8 without BOM".

Comment (wriches) [inline CHIPs/chip-0005.md:180]: JSON should be recommended but not mandatory for off-chain metadata formats, in order to maximise interoperability and reduce the logic required to process multiple formats.

**Review by wriches — CHANGES_REQUESTED** [review]

**Review by wriches — APPROVED** [review]

**Pull request merged**